Fix OutlineInputBorder with BorderRadius.zero is distorted

bleroux opened the pull request:
## Description

This PR fixes an `OutlineInputBorder` rendering bug. When `OutlineInputBorder.borderRadius` is `BorderRadius.zero` and the parent `InputDecorator.labelText` is set and shown on the border, the border is drawn using `Canvas.drawPath` and the path is distorted.

There were some miscalculations mainly related to using a temporary `RRect` with negative border radii.

This PR ~~adds a private `_clampRadius` method to clamp negative border radii and~~ builds the path without calling `Path.addArc` for corners with a zero border radius.
**_Edit_**: `RRect` radii clamping was added to the engine (see https://github.com/flutter/engine/pull/36062 and https://github.com/flutter/engine/pull/36106).

Before this PR (and **before** https://github.com/flutter/engine/pull/36062 + https://github.com/flutter/engine/pull/36106) : 
![before](https://user-images.githubusercontent.com/840911/176601440-69d2da20-8c9f-4ab0-aaaf-deb6aa7f69e7.png)

Before this PR (and **after** https://github.com/flutter/engine/pull/36062 + https://github.com/flutter/engine/pull/36106) : 
![image](https://user-images.githubusercontent.com/840911/190648414-9a7a0228-5cff-4211-ba0b-e8f197ce9a33.png)

After :
![after](https://user-images.githubusercontent.com/840911/176601431-3fb7b2d5-d2bf-4cd3-a631-a28f3394a8bd.png)


## Related Issue

Fixes https://github.com/flutter/flutter/issues/78855
Fixes https://github.com/flutter/flutter/issues/111086

## Tests

Adds 1 test.

bleroux committed:
Fix OutlineInputBorder with BorderRadius.zero is distorted

bleroux commented:
Hey @justinmc!
Do you have any bandwidth to review this PR?
Or maybe pointed out someone who can review it?
Thanks!

bernaferrari commented on packages/flutter/lib/src/material/input_border.dart:
```diff
@@ -415,44 +415,93 @@ class OutlineInputBorder extends InputBorder {
+  }
+
+  // Set negative radii to Radius.zero.
+  RRect _clampToZero(RRect rrect) {
```
clampRadius could be a better name

bernaferrari commented:
If Justin is unavailable, you can try @gspencergoog. He usually enjoys reviewing borders.

gspencergoog commented on packages/flutter/lib/src/material/input_border.dart:
```diff
@@ -439,44 +439,93 @@ class OutlineInputBorder extends InputBorder {
+  }
+
+  // Set negative radii to Radius.zero.
+  RRect _clampRadius(RRect rrect) {
```
I added a draft PR (https://github.com/flutter/engine/pull/36062) to see if anything blows up when we try that.

gspencergoog replied:
In talking to others about this, the consensus was that clamping is just a couple of instructions and we should just do the right thing.  https://github.com/flutter/engine/pull/36062 does clamping in deflate/inflate, and asserts if an `RRect` is created with a negative radius. As soon as the tests pass there, I'll move it to a non-draft PR, and then you can base this change on top of that (which should eliminate some of the code you needed to write).

gspencergoog replied:
Man.  OK, so in order to assert that there aren't any negative radii given to `RRect`, I first need to have an easy way to clamp them, and add that into the Flutter repo code before I can merge the change that causes asserts when `RRect` gets a negative radius, so this is going to be a multi-step process to do the right thing. https://github.com/flutter/engine/pull/36106

bleroux replied:
Great! I will update this PR as soon as https://github.com/flutter/engine/pull/36062 and https://github.com/flutter/engine/pull/36106 will land.

gspencergoog commented:
@bleroux  OK, I've moved the `RRect` clamping through the process, so it's ready for you to update (if you haven't already). `RRect` should now clamp automatically to zero radius on deflate (or inflate with negative values), and will assert when it receives a negative radius.  I'm not sure how much is left to do for this PR after accounting for that, but at the very least the testing is worthwhile.

bleroux commented:
> I'm not sure how much is left to do for this PR after accounting for that, but at the very least the testing is worthwhile.

@gspencergoog That’s great to have RRect clamping at the engine level :+1:  
I had already updated the PR. Today I also updated the PR description with a new screenshot showing that there were remaining miscalculations to fix.

gspencergoog reviewed (approved):
![32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d](https://user-images.githubusercontent.com/8867023/37737237-66e9c9dc-2d10-11e8-8797-13a405eace49.png)

Thanks for fixing this!

justinmc commented:
@bleroux Sorry for missing this PR and thanks @gspencergoog for reviewing!
